### ISSUE TITLE
Serialization problems using Kotlin/JS IR compiler

### ISSUE THREAD

**Issue by robertfmurdock** [kotools/types#431]
## Description

> Bug encountered in version `4.4`.

There appear to be new problems when using javascript in 4.4.1, rending it unusable. Haven't had much time to investigate but here's the stacktrace:

```text
IrLinkageError: Constructor 'NotEmptyListSerializer.<init>' can not be called: No constructor found for symbol '[ File '/Users/robertfmurdock/git/Coupling/libraries/json/src/commonMain/kotlin/com/zegreatrob/coupling/json/JsonPairAssignmentDocument.kt' <- kotools.types.collection/NotEmptyListSerializer.<init>|<init>(kotlinx.serialization.KSerializer<1:0>){}[0] ]'
	at <global>.throwLinkageError(/Users/robertfmurdock/git/Coupling/libraries/json/build/compileSync/js/test/testDevelopmentExecutable/kotlin/js-ir/runtime/unlinked.kt:11)
	at <global>.new Companion(kotlin/Coupling-libraries-json.js:462)
	at <global>.Companion_getInstance(kotlin/Coupling-libraries-json.js:471)
	at <global>.new JsonPinnedCouplingPair(/Users/robertfmurdock/git/Coupling/libraries/json/src/commonMain/kotlin/com/zegreatrob/coupling/json/JsonPairAssignmentDocument.kt:93)
	at <global>.toSerializable(/Users/robertfmurdock/git/Coupling/libraries/json/src/commonMain/kotlin/com/zegreatrob/coupling/json/JsonPairAssignmentDocument.kt:150)
	at <global>.toSerializable(/Users/robertfmurdock/git/Coupling/libraries/model/src/commonMain/kotlin/com/zegreatrob/coupling/model/NotEmptyExtensions.kt:26)
	at <global>.PairDocumentTest$canHandleMilliseconds$lambda(/Users/robertfmurdock/git/Coupling/libraries/json/src/commonMain/kotlin/com/zegreatrob/coupling/json/MiscFuncs.kt:15)
	at <global>.<unknown>(/Users/robertfmurdock/git/Coupling/libraries/json/build/compileSync/js/test/testDevelopmentExecutable/kotlin/src/kotlin/util/Standard.kt:74)
	at <global>.<unknown>(/home/runner/work/testmints/testmints/libraries/standard/src/commonMain/kotlin/com/zegreatrob/testmints/Setup.kt:32)
	at <global>.<unknown>(/home/runner/work/testmints/testmints/libraries/standard/src/commonMain/kotlin/com/zegreatrob/testmints/Setup.kt:98)
	at <global>.StandardMintDispatcher$_get_setup_$lambda_apij2j(/home/runner/work/testmints/testmints/libraries/standard/src/commonMain/kotlin/com/zegreatrob/testmints/StandardMintDispatcher.kt:7)
	at <global>.checkedInvoke(/home/runner/work/testmints/testmints/libraries/standard/src/commonMain/kotlin/com/zegreatrob/testmints/Setup.kt:96)
	at <global>.runTest(/home/runner/work/testmints/testmints/libraries/standard/src/commonMain/kotlin/com/zegreatrob/testmints/Setup.kt:25)
	at <global>.<unknown>(/home/runner/work/testmints/testmints/libraries/standard/src/commonMain/kotlin/com/zegreatrob/testmints/Setup.kt:14)
	at Exercise.protoOf.verify_otwzpa(/home/runner/work/testmints/testmints/libraries/standard/src/commonMain/kotlin/com/zegreatrob/testmints/Exercise.kt:5)
	at PairDocumentTest.protoOf.canHandleMilliseconds_xgna86(/Users/robertfmurdock/git/Coupling/libraries/json/src/jsTest/kotlin/com/zegreatrob/coupling/json/PairDocumentTest.kt:12)
	at <global>.fn(kotlin/Coupling-libraries-json-test.js:316)
	at Context.<anonymous>(/Users/robertfmurdock/git/Coupling/build/js/packages_imported/kotlin-test-js-runner/src/KotlinTestTeamCityConsoleAdapter.ts:70)
	at <global>.processImmediate(node:internal/timers:478)
```

## Checklist

- Fix the serialization of the following types:
  - [x] `AnyInt`
  - [x] `PositiveInt`
  - [x] `NegativeInt`
  - [x] `StrictlyPositiveInt`
  - [x] `StrictlyNegativeInt`
  - [x] `ZeroInt`
  - [x] `StrictlyPositiveDouble`
  - [x] `NotBlankString`
  - [x] `EmailAddress`
  - [x] `NotEmptyList`
  - [x] `NotEmptySet`
  - [x] `NotEmptyMap`
- [x] Update the unreleased changelog.
- [x] Apply these changes to the `release/4.4.2` branch.

**Comment by LVMVRQUXL** [comment]
Thank you for reporting this bug @robertfmurdock.

It seems related to the serialization of the `NotEmptyList` type, but I don't clearly understand the context of this error.
How do you use this type on Kotlin/JS?

**Comment by robertfmurdock** [comment]
example of usage:

`
@Serializable
data class JsonPairAssignmentDocument(
    val id: String,
    val date: Instant,
    val pairs: NotEmptyList<JsonPinnedCouplingPair>,
    val discordMessageId: String? = null,
    val slackMessageId: String? = null,
)
`

So nothing special, just importing and using in an object used by kotlin serialization.

**Comment by LVMVRQUXL** [comment]
@robertfmurdock Is `JsonPinnedCouplingPair` serializable?

**Comment by robertfmurdock** [comment]
Yes indeed

**Comment by LVMVRQUXL** [comment]
@robertfmurdock I managed to reproduce this bug locally and found the problem.
Because we've changed the visibility of the `NotEmptyListSerializer` type from `internal` (4.3.1) to `private` (4.4.0) in commit 86ead9f, the compiler can't find it for serializing the `NotEmptyList` type.
So the solution is to make all serializers internally visible.

**Comment by LVMVRQUXL** [comment]
The fix of this bug is available in the [4.4.2 release](https://github.com/kotools/types/releases/tag/4.4.2).

**Comment by LVMVRQUXL** [comment]
@robertfmurdock The fix of this bug is available in the [4.4.2 release](https://github.com/kotools/types/releases/tag/4.4.2).
